### PR TITLE
Add spent outputs to PrecomputedTransactionData

### PR DESCRIPTION
**Pull request by EyeOfPython**
This is required for the BIP341 sighash.

### CODE REVIEW

File: src/test/transaction_tests.cpp [hunk 1]
```diff
@@ -887,8 +898,8 @@ BOOST_AUTO_TEST_CASE(txsize_activation_test) {
     CTransaction minTx;
     TxValidationState state;
 
-    BOOST_CHECK(!ContextualCheckTransaction(
-        params, minTx, state, 0, 5678, 1234));
+    BOOST_CHECK(
```
Comment (schancel) [inline src/test/transaction_tests.cpp:901]: Would it be possibly to patch add the changes without these formatting adjustments?

File: src/test/transaction_tests.cpp [hunk 2]
```diff
@@ -438,12 +445,16 @@ BOOST_AUTO_TEST_CASE(test_big_transaction) {
...
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(mtx.vin.size());
     for (size_t i = 0; i < mtx.vin.size(); i++) {
         CTxOut out;
         out.nValue = 1000 * SATOSHI;
         out.scriptPubKey = scriptPubKey;
+        spent_outputs.push_back(out);
         coins.emplace_back(std::move(out), 1, false);
     }
+    PrecomputedTransactionData txdata(tx, std::move(spent_outputs));
```
Comment (schancel) [inline src/test/transaction_tests.cpp:457]: What's this for?

File: src/test/script_p2sh_tests.cpp [hunk 3]
```diff
@@ -110,15 +110,14 @@ BOOST_AUTO_TEST_CASE(sign) {
...
+        PrecomputedTransactionData txdata(tx, {txFrom.vout[i]});
         for (int j = 0; j < 8; j++) {
             CScript sigSave = txTo[i].vin[0].scriptSig;
             txTo[i].vin[0].scriptSig = txTo[j].vin[0].scriptSig;
             bool sigOK =
                 CScriptCheck(txFrom.vout[txTo[i].vin[0].prevout.GetN()],
                              CTransaction(txTo[i]), 0,
-                             SCRIPT_ENABLE_SIGHASH_FORKID,
-                             false, txdata)();
+                             SCRIPT_ENABLE_SIGHASH_FORKID, false, txdata)();
```
Comment (schancel) [inline src/test/script_p2sh_tests.cpp:120]: Formatting?

**Review by schancel — COMMENTED** [review]
Makes sense.

File: src/test/transaction_tests.cpp (hunk 1 above)
Comment (EyeOfPython) [reply src/test/transaction_tests.cpp:901]: Will do.

File: src/test/transaction_tests.cpp (hunk 2 above)
Comment (EyeOfPython) [reply src/test/transaction_tests.cpp:457]: Once we add tx tests using the new BIP341 sighash, txdata will have the correct `spent_outputs`.

Comment (EyeOfPython) [reply src/test/transaction_tests.cpp:457]: Better make this correct, or people will be very confused in the future.

Comment (schancel) [reply src/test/transaction_tests.cpp:457]: It just doesn't seem to be used right now. Why not add it later?

Comment (schancel) [reply src/test/transaction_tests.cpp:457]: Ah. Nevermind. I did not see that the stuff above was part of this test.

File: src/script/bitcoinconsensus.cpp [hunk 4]
```diff
@@ -99,7 +99,13 @@ static int verify_script(const uint8_t *scriptPubKey,
...
         set_error(err, bitcoinconsensus_ERR_OK);
 
-        PrecomputedTransactionData txdata(tx);
+        // Note: The new BIP341 sighash requires knowing all CTxOut of the tx's
+        // inputs. We don't have access to this here yet, so verification of
+        // transactions that rely on the outputs will fail.
+        // Eventually we should add a
+        // bitcoinconsensus_verify_script_with_spent_outputs function to address
+        // this.
+        PrecomputedTransactionData txdata(tx, {});
```
Comment (schancel) [inline src/script/bitcoinconsensus.cpp:108]: I'd make sure you had access to the coinsview or whatever you need here first. That way there can't be a mistake later.

**Review by schancel — APPROVED** [review]
Out of band conversation about piping the coinsview into `verify_script` before doing this.

File: src/script/bitcoinconsensus.cpp [hunk 5]
```diff
@@ -99,7 +99,13 @@ static int verify_script(const uint8_t *scriptPubKey,
         // Regardless of the verification result, the tx did not error.
         set_error(err, bitcoinconsensus_ERR_OK);
 
-        PrecomputedTransactionData txdata(tx);
+        // Note: The new BIP341 sighash requires knowing all CTxOut of the tx's
```
Comment (schancel) [inline src/script/bitcoinconsensus.cpp:102]: My feedback is basically, if you do the piping first, nobody will be able to make verification code that doesn't work based on this.

File: src/validation.cpp [hunk 6]
```diff
@@ -1789,10 +1791,11 @@ bool CChainState::ConnectBlock(const CBlock &block, BlockValidationState &state,
         int nSigChecksRet;
         TxValidationState tx_state;
         if (fScriptChecks &&
-            !CheckInputScripts(tx, tx_state, view, flags, fCacheResults,
-                               fCacheResults, PrecomputedTransactionData(tx),
-                               nSigChecksRet, nSigChecksTxLimiters[txIndex],
-                               &nSigChecksBlockLimiter, &vChecks)) {
+            !CheckInputScripts(
```
Comment (schancel) [inline src/validation.cpp:1794]: Is the view here going to be able to handle transactions that are chained in the same block?

File: src/script/bitcoinconsensus.cpp [hunk 7]
```diff
@@ -128,9 +130,10 @@ int bitcoinconsensus_verify_script(const uint8_t *scriptPubKey,
         flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_WITNESS_DEPRECATED) {
         return set_error(err, bitcoinconsensus_ERR_AMOUNT_REQUIRED);
     }
+    // TODO: Return failure if BIP341 flag is set.
```
Comment (schancel) [inline src/script/bitcoinconsensus.cpp:133]: Agree

**Review by schancel — CHANGES_REQUESTED** [review]
Looks sane, should add the assert in `bitcoinconsensus_verify_script` now, and I also had a question about the ConnectBlock's view.

File: src/validation.cpp (hunk 6 above)
Comment (EyeOfPython) [reply src/validation.cpp:1794]: I'm pretty sure CCoinsView handles it. Not sure how to test it at this moment

Comment (schancel) [reply src/validation.cpp:1794]: Answered my question. Up on line 1704 the outputs are all added to the view are all added to be able to check the inputs for CTOR.

Comment (EyeOfPython) [reply src/validation.cpp:1794]: Shammah [Not asking for crypto], [3/18/21, 8:42:24 PM]:
> line 1704 answers my question
> The outputs are all added to the view before checking anything

**Review by schancel — APPROVED** [review]

**Pull request merged**